### ISSUE TITLE
Move retention apply to the beginning of Compactor loop 

### ISSUE THREAD

**Issue by bwplotka** [thanos-io/thanos#1708]
We apply retention at the end, we probably should move it before any compaction. It was by design to make sure lower resolution can be downsampled, but TBH it does not really make much sense. If you configure the retention in a way that raw goes away too quickly this can hit you anyway. So I cannot see the blocker now (: 

We can save a lot especially if retention is desired to be e.g 3 months but we left the data without compactor for 6 months (: We can avoid unnecessary compactions and downsampling.

AC: 
* Move retention applies to the beginning of Thanos compact loop.

**Comment by GiedriusS** [comment]
>  If you configure the retention in a way that raw goes away too quickly this can hit you anyway.

That's true but I would really love to make Thanos Compact scream loudly if this happens as it indicates some kind of error, no? Especially if downsampling is enabled but the retention policies prevent us from downsampling data.

**Comment by bwplotka** [comment]
I think we might need to make it a bit more complex.

Let's consider this case:
* raw retention is 1month
* downsampled retention is 2y
* compactor is not running for 2 months.

Now what will happen with this change is that we will have a gap for downsampled data. between `[2mo, 1mo]` as we remove raw data that was dependent on. 

I think we could improve the flow only for case where raw = res5m = res1h retention is equal. In this case, removal on the start might make sense. 

Or in other words: We can only remove up to res1h retention at the start. Does it make sense? Other opinions are welcome (:

**Comment by GiedriusS** [comment]
Reopening since there is a new PR which tries to address this.

**Comment by kakkoyun** [comment]
Still ongoing. Let's keep it fresh for a little more.

**Comment by GiedriusS** [comment]
In progress.

**Comment by AllenzhLi** [comment]
Maybe this issue should be more attention.

**Comment by heyitsmdr** [comment]
Is this still on-going by any chance?

**Comment by yeya24** [comment]
This can also address the problem mentioned in https://github.com/thanos-io/thanos/pull/4406
